pytorch_lightning/plugins/training_type/deepspeed.py

```diff
@@ -0,0 +1,226 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import logging
+import os
+from pathlib import Path
+from types import SimpleNamespace
+from typing import Any, Callable, List, Optional, Tuple, Union
+
+import torch
+
+from pytorch_lightning.core.lightning import LightningModule
+from pytorch_lightning.overrides.base import _LightningModuleWrapperBase
+from pytorch_lightning.plugins.environments.cluster_environment import ClusterEnvironment
+from pytorch_lightning.plugins.training_type.ddp import DDPPlugin
+from pytorch_lightning.utilities import AMPType
+from pytorch_lightning.utilities.apply_func import apply_to_collection
+from pytorch_lightning.utilities.distributed import rank_zero_only
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from pytorch_lightning.utilities.imports import _DEEPSPEED_AVAILABLE
+from pytorch_lightning.utilities.seed import seed_everything
+
+if _DEEPSPEED_AVAILABLE:
+    import deepspeed
+else:
+    deepspeed = None
+
+
+class LightningDeepSpeedModule(_LightningModuleWrapperBase):
+
+    def __init__(self, pl_module: LightningModule, precision: int):
+        super().__init__(pl_module)
+        self.module = pl_module
+        self.precision = precision
+
+    def forward(self, *inputs, **kwargs):
+        if self.precision == 16:
+            inputs = self._move_float_tensors_to_half(inputs)
+
+        return super().forward(*inputs, **kwargs)
+
+    def _move_float_tensors_to_half(self, batch: Any):
+
+        def batch_to(data):
+            return data.half()
+
+        batch = apply_to_collection(batch, (torch.FloatTensor, torch.cuda.FloatTensor), function=batch_to)
```

Comment:
Can we please not use specific tensor types here, but instead use torch.Tensor here and include the dtype check into the `batch_to` funtion? Not sure if this also supports ROCm for instance.

Also what about long-Tensors or double-tensors?

Reply:
torch.Tensor doesn't define `half()`. you can only convert float tensors to float16. I think it's important to be explicit to the tensors which are supported via the `half` function, so we could make it even more general by using a `hasattr` check, but I think thats overkill

Reply:
I'm not getting you @SeanNaren. This works:

```python
>>> x  = torch.tensor([1, 2, 3], dtype=torch.float)
>>> x.to(torch.half)
tensor([1., 2., 3.], dtype=torch.float16)

>>> x  = torch.tensor([1, 2, 3], dtype=torch.long)
>>> x.to(torch.half)
tensor([1., 2., 3.], dtype=torch.float16)
```

Reply:
this turns int/long tensors into floating point tensors which is incorrect... imagine if you have a set of one hot vectors used for embedding mappings, you want to keep them as int tensors

Reply:
Yes makes sense. I was referring to having batch_to as:

```python
if data.dtype == torch.float:
    data.to(torch.half)
return data
```